### DISCUSSION TITLE
Differences in `.clangd` vs `compile_commands.json` in determining the path to the system headers

### DISCUSSION THREAD

**Discussion by palapapa** [clangd/clangd#1985] [Q&A] [answered]
With a `.clangd` file, you can specify the `Compiler` property to make clangd extract its include paths. Clangd also does the same automatically with a `compile_commands.json`. It seems like these two files have an overlap in their features. How do they affect each other in regards to determining the include paths?

In general, how do `.clangd` and `compile_commands.json` differ? Why wouldn't I always use `.clangd` to configure clangd?

**Comment by i-ky** [comment, marked as answer]
> How do they affect each other in regards to determining the include paths?

I believe that `Compiler` from `.clangd` file can override the compiler executable specified in `compile_commands.json`. This may be useful if, for example, "real" compiler from `compile_commands.json` does not behave like GCC or clang, has implicit defines, etc. `Compiler` allows you to point clangd to a script that will fake the compiler output so that clangd can extract include paths, defines, etc.

> In general, how do `.clangd` and `compile_commands.json` differ? Why wouldn't I always use `.clangd` to configure clangd?

The biggest difference is that `compile_commands.json` is typically generated by build system (e.g. CMake) or special tools (e.g. Bear). `.clangd` is typically hand-written.

I would always prefer `compile_commands.json`, because it keeps clangd knowledge of project's compilation options in sync with build system. But I would still have `.clangd` to add extra compiler warning options, enable clang-tidy checks, etc.

**Reply by HighCommander4** [reply]
> I believe that `Compiler` from `.clangd` file can override the compiler executable specified in `compile_commands.json`. This may be useful if, for example, "real" compiler from `compile_commands.json` does not behave like GCC or clang, has implicit defines, etc. `Compiler` allows you to point clangd to a script that will fake the compiler output so that clangd can extract include paths, defines, etc.

Yep, that's right.

Another use case for `Compiler` is to specify the compiler (usually for the purpose of querying system include directories using `--query-driver`) for projects that don't have a `compile_commands.json` file.

> > In general, how do `.clangd` and `compile_commands.json` differ? Why wouldn't I always use `.clangd` to configure clangd?

To add to what @i-ky wrote, the capabilities that `compile_commands.json` has which `.clangd` doesn't are:

 * Providing clangd with a list of source files in the project (namely, the entries in the `compile_commands.json`) so clangd knows what to index. This is a requirement for clangd's background indexer, which powers many cross translation unit features like finding references across the project.
 * Specifying different flags for different files. By comparison, `.clangd` only gives you directory-level granularity. (I suppose you could get file-level granularity with `PathMatch` but that seems a bit tedious.)
